torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Any, Dict, Optional
+
+import torch
+
+_WORKER_ID = "worker_id"
+_FETCHER_STATE = "fetcher_state"
+_FETCHER_ENDED = "fetcher_ended"
+_DATASET_STATE = "dataset_state"
+_DATASET_ITER_STATE = "dataset_iter_state"
+
+
+def _flatten(nested_data, key_lineage=()):
+    if nested_data is None:
+        return None
+    data = {}
+    if isinstance(nested_data, dict):
+        for key, value in nested_data.items():
+            flat = _flatten(value, key_lineage + (key,))
+            data.update(flat)
```

Comment:
I still think we could add dataclass support but let's do it in a subsequent diff. I am unsure if there is significant overhead of dataclasses.asdict, since it appears they're making copies, eg 
![Uploading image.png…]()